typescript/common-resolvers/src/import/format-coco/index.ts

```diff
@@ -0,0 +1,174 @@
+import JSZip from "jszip";
+import path from "path";
+import { v4 as uuidv4 } from "uuid";
+
+import type { UploadTargetHttp } from "@labelflow/graphql-types";
+
+import { ImportFunction } from "../types";
+import imageResolvers from "../../image";
+import labelClassResolvers from "../../label-class";
+import labelResolvers from "../../label";
+import { CocoDataset } from "../../export/format-coco/coco-core/types";
+import { Context } from "../../types";
+import { convertCocoSegmentationToLabel } from "./converters";
+
+const uploadImage = async (
+  file: JSZip.JSZipObject,
+  name: string,
+  datasetId: string,
+  { repository }: Context
+) => {
+  const fileBlob = await file.async("blob", () => {});
+  const uploadTarget = await repository.upload.getUploadTarget(uuidv4());
+  if (!(uploadTarget as UploadTargetHttp)?.downloadUrl) {
+    throw new Error("Can't direct upload this image.");
+  }
+  repository.upload.put(
+    (uploadTarget as UploadTargetHttp)?.downloadUrl,
+    fileBlob
+  );
+  return await imageResolvers.Mutation.createImage(
+    null,
+    {
+      data: {
+        url: (uploadTarget as UploadTargetHttp)?.downloadUrl,
+        name,
+        datasetId,
+      },
+    },
+    { repository }
+  );
+};
+
+export const importCoco: ImportFunction = async (
+  blob,
+  datasetId,
+  { repository },
+  options
+) => {
+  let annotationBlob = blob;
+  if (!options?.annotationsOnly) {
+    const zip = await JSZip.loadAsync(blob);
+    const annotationsFilesJSZip = zip.filter(
+      (relativePath) => path.extname(relativePath) === ".json"
+    );
+    if (annotationsFilesJSZip.length === 0) {
+      throw new Error("No COCO annotation file was found in the zip file.");
+    }
+    if (annotationsFilesJSZip.length > 1) {
+      throw new Error(
+        "More than one COCO annotation file was found in the zip file."
+      );
+    }
+    const annotationFile: CocoDataset = JSON.parse(
+      await annotationsFilesJSZip[0].async("string", () => {})
+    );
+    annotationBlob = await annotationsFilesJSZip[0].async("blob", () => {});
+    const imageFiles = zip.filter((relativePath) =>
+      path.dirname(relativePath).endsWith("images")
+    );
+    const imageNameToFile = imageFiles.reduce(
+      (imageNameToFileCurrent, imageFile) => {
+        imageNameToFileCurrent.set(path.basename(imageFile.name), imageFile);
+        return imageNameToFileCurrent;
+      },
+      new Map()
+    );
+    // Manage coco images => labelflow images
+    const images = await repository.image.list({ datasetId });
+    const imagesCoco = annotationFile.images.filter(
+      (imageCoco) =>
+        !images.find(
+          (image) =>
+            image.name ===
+            imageCoco.file_name.replace(path.extname(imageCoco.file_name), "")
+        )
+    );
+    // eslint-disable-next-line no-restricted-syntax
+    for (const imageCoco of imagesCoco) {
```

Comment:
Why using a `for` here?

Reply:
I need to await image creation sequentially, and I didn't want to import bluebird for that...  Maybe you'll have an alternative option?

Reply:
Why does it need to be sequential?

Reply:
To ensure that `export ∘ import = Identity`